nemo/collections/llm/gpt/data/utils.py

```diff
@@ -799,60 +802,97 @@ def _preprocess(
-    2. Tokenize the chat message by applying the chat template.
-    3. Calculate the mask for the assistant tokens.
+    if not hasattr(tokenizer.tokenizer, "apply_chat_template"):
+        raise ValueError("Cannot apply chat template with tokenizer that is not a HuggingFace AutoTokenizer")
+
+    tools = None
+    if isinstance(source, dict):
+        tools = source.get("tools") or tool_schemas
+
+        if source.get("conversations"):
+            # Detect if Nemo {"conversations": [{"from": "User/Assistant", "value": ""}]}
+            # Convert to HuggingFace chat template [{"role": "stystem/user/assistant", "content": ""}]
+            chat = [{"role": convo["from"].lower(), "content": convo["value"]} for convo in source["conversations"]]
+            if source.get("system"):
+                chat.insert(0, {"role": "system", "content": source["system"]})
+
+        elif source.get("messages"):
+            # HuggingFace chat template {"messages": [{"role": "system/user/assistant", "content": ""}]}
+            chat = source.get("messages")
+    else:
+        chat = source
 
-    Args:
-        hf_chat_dict: dict, the chat message dict from HuggingFace tokenizer.
-        tokenizer: TokenizerSpec, the tokenizer object.
-    """
-    # Use HuggingFace tokenizer to apply the chat template.
-    template_has_generation_kwd = (
-        '{% generation %}' in tokenizer.tokenizer.chat_template
-    )  # assistant mask only works if chat template has generation keyword
-    tokens = tokenizer.tokenizer.apply_chat_template(
-        hf_chat_dict['messages'],
+    # assistant mask only works if chat template has generation keyword
+    template_has_generation_kwd = GENERATION_REGEX.search(tokenizer.tokenizer.chat_template) is not None
+
+    tokenized_chat = tokenizer.tokenizer.apply_chat_template(
+        chat,
+        tools=tools,
+        tokenize=True,
         return_dict=True,
         return_assistant_tokens_mask=template_has_generation_kwd,
-        return_tensors='pt',
     )
-    input_ids = tokens['input_ids'][0]
+
+    # Choose the last conversation as answer other history are context by finding the last masked token
+    # which indicates end of context and beginning of answer
+    input_ids = tokenized_chat.get("input_ids")
     if template_has_generation_kwd:
-        mask = torch.tensor(tokens['assistant_masks']).to(bool)
+        mask = tokenized_chat['assistant_masks']
     else:
-        mask = torch.ones_like(input_ids)
-    return dict(input_ids=input_ids, mask=mask)
+        mask = [1] * len(input_ids)
+
+    if tokenizer.eos_id and input_ids[-1] != tokenizer.eos_id:
+        input_ids += [tokenizer.eos_id]
+        mask += [1]
+
+    if 0 in mask:
+        # traverse the list backward for first occurrence of masked token
+        context_end_idx = len(mask) - mask[::-1].index(0)
+    else:
+        context_end_idx = len(mask)
+
+    context_ids = input_ids[:context_end_idx]
```

Comment:
what do you need the `context_ids` and `answer_ids` for in SFT? I think SFT works without them too

Reply:
It maintains consistency in the interface in outputs and inputs.